code/harry_potter_store/src/store.py

```diff
@@ -0,0 +1,85 @@
+from copy import deepcopy
+from typing import Dict
+
+
+class Store:
+    book_types = ["1", "2", "3", "4", "5"]
+
+    def __init__(self, book_price=8):
+        self.book_price = book_price
+        pass
+
+    def _validate(self, to_buy_books: Dict[str, int]):
+        for book_type, amount in to_buy_books.items():
+            if book_type not in Store.book_types:
+                raise ValueError("Book type %s must be in %s".format(book_type, Store.book_types))
+            if amount <= 0:
+                raise ValueError("Book amount %s for type %s must be > 0".format(amount, book_type))
+
+    def _clean(self, to_buy_books: Dict[str, int]) -> dict:
+        return dict(filter(lambda item: item[1] > 0 and item[0] in Store.book_types, to_buy_books.items()))
+
+    def _decrement(self, to_buy_books: Dict[str, int], count: int) -> dict:
+        first_part = {book: amount - 1 for book, amount in list(to_buy_books.items())[:count]}
+        second_part = dict(list(to_buy_books.items())[count:])
+        first_part.update(second_part)
+        return self._clean(first_part)
+
+    def _discount_wrapper(self, number, discount):
+        return lambda to_buy_books, price, predicate: \
+            self._discount(to_buy_books, price, predicate, number, discount)
+
+    def _discount(self, to_buy_books, price, predicate, number, discount):
+        if predicate(len(to_buy_books), number):
+            price += number * self.book_price * (1 - discount)
+            to_buy_books = self._decrement(to_buy_books, number)
+        return to_buy_books, price
+
+    """
+    algorithm logic: we have discounts for several books. We give names for it: 3, 2, 1
+    and we are doing a full sort out
+    we calculate a price just with discount for 3 books
+    next for 2 books
+    next for 1 books
+    and next we calculate a price with discount for 3, 2, 1 books
+    next for 2, 1 books
```

Comment:
why? =) Is there a case where it's needed?

Reply:
when we have {"1": 2, "2": 1} then we have discount for 2 books and full price for another one

Reply:
This doesn't answer the question. Why do you need a check for "2, 1" if you can have a check for "3,2,1"?

Reply:
you example {"1": 2, "2": 1, "3": 2, "4": 1, "5": 2} has length is 5 but best answer discount was 4
why we cannot have such case for example with length is 3 ?

Reply:
This is covered by the first part where you iterate with `>=` predicate.

Reply:
you are right. I see it now.
i changed condition from `i != 0` to `i > 1` to doesn't executing extra computation and also I flipped over discount functions for that